# Parnter Business Modeling - Uber

## Assignment

<b>Scenario 1: It is going to be a huge Saturday and there will need to be many more cars on the road than last week. In order to get drivers to go online, we're assessing the following two bonus options in terms of cost:</b>

- Option 1: \$50 for each driver that is online at least 8 hours, accepts 90% of requests, completes 10 trips, and has a rating of 4.7 or better during the time frame;
- Option 2: $4/trip for all drivers who complete 12 trips, and have a 4.7 or better rating.
Using the dataset provided and given Scenario 1, provide answers to the questions below:

- How much would the total bonus payout be with Option 1?
- How much would the total bonus payout be with Option 2?
- How many drivers would qualify for a bonus under Option 1 but not under Option 2?
- What percentages of drivers online completed less than 10 trips, had an acceptance rate of less than 90%, and had a rating of 4.7 or higher?

<b>Scenario 2: A taxi driver currently generates $200 per day in fares (before expenses), works six days a week, takes three weeks off, and has the following expenses:</b>

- Gas - \$200 per week
- Insurance - \$400 per month
- Vehicle rent (by the week) - \$500

The driver doesn't pay gas and rent expenses on off weeks.

Now, let's assume that the same driver would buy a Town Car and partner with Uber. If he does, his gas expenses would go up by 5%, his insurance expense would decrease by 20%, and he would no longer be renting a vehicle. However, he would need to buy a car. The driver would still take three weeks off per year.

Given Scenario 2, provide answers to the questions below:

- How much money (after expenses) does the taxi driver make per year without partnering with Uber?
- You are convincing the same driver above to buy a Town Car and partner with Uber. Assuming the new car is 40,000 USD, how much would the driver's gross fares need to increase per week to fully pay for the car in year 1 and maintain the same yearly profit margin as before?


## Data Analysis and Cleaning

In [1]:
# Import necessary libraries
import pandas as pd
import numpy as np

The dataset is given in the file `dataset_2.csv`. The data come from rides on a busy Saturday and are aggregated on a driver level. The dataset provides information on the total number of completed trips, the driver's acceptance rate, the total number of hours on duty, and the average rating.

In [2]:
# Read csv file as df and get the first 5 rows
df = pd.read_csv('dataset_2.csv')

df.head()


Name  Trips Completed Accept Rate  Supply Hours  Rating
0    Abdul                1        100%             3     4.8
1  Abraham               12         83%             5     4.7
2  Adelina                1        100%             2     4.7
3   Akilah                1        100%             2     4.9
4     Alec               21         76%            11     5.0

In [3]:
# Check if there's missing value
df.isna().sum()

Name               0
Trips Completed    0
Accept Rate        0
Supply Hours       0
Rating             0
dtype: int64

In [4]:
df.describe()

Trips Completed  Supply Hours      Rating
count       119.000000    119.000000  119.000000
mean         11.941176      7.084034    4.698571
std           5.857480      2.794000    0.406876
min           1.000000      1.000000    1.600000
25%           8.000000      5.000000    4.600000
50%          12.000000      7.000000    4.800000
75%          15.000000      9.000000    4.900000
max          31.000000     11.000000    5.000000

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119 entries, 0 to 118
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             119 non-null    object 
 1   Trips Completed  119 non-null    int64  
 2   Accept Rate      119 non-null    object 
 3   Supply Hours     119 non-null    int64  
 4   Rating           119 non-null    float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.8+ KB


In [6]:
# Check if there's duplicates
df.drop_duplicates()

Name  Trips Completed Accept Rate  Supply Hours  Rating
0       Abdul                1        100%             3     4.8
1     Abraham               12         83%             5     4.7
2     Adelina                1        100%             2     4.7
3      Akilah                1        100%             2     4.9
4        Alec               21         76%            11     5.0
..        ...              ...         ...           ...     ...
114    Virgen                4        100%             6     4.5
115      Yang                7         71%             2     4.5
116  Yessenia                8         88%             5     5.0
117    Yukiko                9         78%             6     4.5
118       Zoe               21         86%             9     4.9

[119 rows x 5 columns]

There is no duplicates.

Since the `Accept Rate` is in type object we won't be able to work on it. Let's change it to integer

In [7]:
# Remove % from the value and chaneg the datatype to integer
df['Accept Rate'] = df['Accept Rate'].str.replace('%', '').astype('int')

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119 entries, 0 to 118
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             119 non-null    object 
 1   Trips Completed  119 non-null    int64  
 2   Accept Rate      119 non-null    int64  
 3   Supply Hours     119 non-null    int64  
 4   Rating           119 non-null    float64
dtypes: float64(1), int64(3), object(1)
memory usage: 4.8+ KB


Total number of drivers is 119, with no duplicates or missing data. Let's move on to solving our questions. 

## Scenario 1

Uber providing two bonus options in terms of cost:

- Option 1: $50 for each driver that is online at least 8 hours, accepts 90% of requests, completes 10 trips, and has a rating of 4.7 or better during the time frame;

<b>How much would the total bonus payout be with Option 1?</b>

In [9]:
option1_df = df[(df['Supply Hours'] >= 8) & (df['Accept Rate'] >= 90) & (df['Trips Completed'] >= 10) & (df['Rating'] >= 4.7)]
option1_count = option1_df['Name'].count()

print(f"Total Bonus Payout for Option 1: ${option1_count * 50}")

Total Bonus Payout for Option 1: $1050


- Option 2: $4/trip for all drivers who complete 12 trips, and have a 4.7 or better rating. Using the dataset provided and given Scenario 1, provide answers to the questions below:

<b>How much would the total bonus payout be with Option 2?</b>

In [10]:
option2_df = df[(df['Trips Completed'] >= 12) & (df['Rating']>=4.7)]
option2_sum = option2_df['Trips Completed'].sum()

print(f"Total Bonus Payout for Option 2: ${option2_sum * 4}")

Total Bonus Payout for Option 2: $2976


<b>How many drivers would qualify for a bonus under Option 1 but not under Option 2?</b>

In [11]:
filtered_df = df[(df['Name'].isin(option1_df['Name'])) & (~df['Name'].isin(option2_df['Name']))]
op1_notop2 = filtered_df['Name'].count()

print(f"Number of Drivers Qualify for Option 1 but not for Option 2: {op1_notop2} Drivers")

Number of Drivers Qualify for Option 1 but not for Option 2: 2 Drivers


In [12]:
filtered_df

Name  Trips Completed  Accept Rate  Supply Hours  Rating
88     Oren               11           91             9     4.8
92  Phyllis               10           90             8     4.8

<b>What percentages of drivers online completed less than 10 trips, had an acceptance rate of less than 90%, and had a rating of 4.7 or higher?

In [13]:
less10trip_df = df[(df['Trips Completed'] < 10) & (df['Accept Rate'] < 90) & (df['Rating'] >= 4.7)]
less10trip_numbers = less10trip_df['Name'].count()

print(f"Percentages of drivers: {round(less10trip_numbers/df['Name'].count() * 100, 2)}%")

Percentages of drivers: 10.92%


## Scenario 2

A taxi driver currently generates $200 per day in fares (before expenses), works six days a week, takes three weeks off, and has the following expenses:

- Gas - \$200 per week
- Insurance - \$400 per month
- Vehicle rent (by the week) - $500

The driver doesn't pay gas and rent expenses on off weeks.

Now, let's assume that the same driver would buy a Town Car and partner with Uber. If he does, his gas expenses would go up by 5%, his insurance expense would decrease by 20%, and he would no longer be renting a vehicle. However, he would need to buy a car. The driver would still take three weeks off per year.

<b>How much money (after expenses) does the taxi driver make per year without partnering with Uber?

In [35]:
# Set parameters to avoid hardcoding

total_week_in_year = 52
working_days_in_week = 6
weeks_off = 3
total_months_in_year = 12
fare_per_day = 200

# Expenses
gas_per_week = 200
insurance_per_month = 400
vehicle_rent_per_week = 500

In [24]:
total_revenue = (total_week_in_year - weeks_off) * working_days_in_week * fare_per_day
print(f"${total_revenue}")

$58800


In [28]:
total_cost = (total_week_in_year - weeks_off) * (gas_per_week + vehicle_rent_per_week) + total_months_in_year * insurance_per_month
print(f"${total_cost}")

$39100


In [30]:
total_profit = total_revenue - total_cost
print(f"${total_profit}")

$19700


<b>You are convincing the same driver above to buy a Town Car and partner with Uber. Assuming the new car is 40,000 USD, how much would the driver's gross fares need to increase per week to fully pay for the car in year 1 and maintain the same yearly profit margin as before?

In [38]:
# New costs parameters
# Gas expense go up by 5% per week
# Insurance drop by 20% (paying only 80% from current amount) per month
new_car_cost = 40000
new_gas_per_week = 200 * 1.05
new_insurance_per_month = 400 * 0.8

In [34]:
# Cost per for year 1 partnering with Uber buying the Town Car
# Add the cost of buying a town car (40000)
uber_partner_cost = new_car_cost + new_gas_per_week * (total_week_in_year - weeks_off) + new_insurance_per_month * total_months_in_year
uber_partner_cost

54130.0

In [53]:
gross_fare_increase_needed = (uber_partner_cost + total_profit - total_revenue) / (total_week_in_year - weeks_off)
print(f"Driver's Gross Fares needed to Increase: ${round(gross_fare_increase_needed - fare_per_day, 2)}")
print(f"Driver's Gross Fare needed: ${round(gross_fare_increase_needed, 2)}")

Driver's Gross Fares needed to Increase: $106.73
Driver's Gross Fare needed: $306.73
